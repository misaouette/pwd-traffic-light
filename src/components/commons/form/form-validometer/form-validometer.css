:host {
  display: block;
  font-size: 0.8rem;
}
:host([vspace-small]) {
  margin-top: var(--vmargin-small);
}

.criteria {
  display: flex;
}

.icon {
  --icon-size: 0.9rem;
  background-size: var(--icon-size);
  height: var(--icon-size);
  width: var(--icon-size);
  margin-right: 0.2rem;
  margin-top: 0.1rem;
}
.invalid .icon {
  background-image: url(./assets/validation/invalid.png);
}
.valid .icon {
  background-image: url(./assets/validation/valid.png);
}

mark {
  background-color: transparent;
}
.invalid mark {
  color: #d91e18;
}
.valid mark {
  color: #00aa55;
}
