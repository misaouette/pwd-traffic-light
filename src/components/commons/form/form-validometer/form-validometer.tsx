import { Component, Host, Prop, Watch, h } from '@stencil/core';

/// <reference path="../../../../types/Validation.ts" />

@Component({
  tag: 'form-validometer',
  styleUrl: 'form-validometer.css',
  shadow: true,
})
export class FormValidometer {
  @Prop() criteriaList!: Validation.Criteria[];
  @Prop({ reflect: true }) vspaceSmall: boolean;

  @Watch('criteriaList')
  validateCriteriaList(newCriteriaList: Validation.Criteria[]) {
    let invalidCriteria;

    invalidCriteria = newCriteriaList.find(({ isValid, message }) => isValid !== undefined && !message.includes('<mark>'));
    if (invalidCriteria) {
      throw new Error(`message in [\{message, isValid\}] must contain an opening <mark> tag : (${invalidCriteria.message})`);
    }

    invalidCriteria = newCriteriaList.find(({ isValid, message }) => isValid !== undefined && !message.includes('</mark>'));
    if (invalidCriteria) {
      throw new Error('message in [{message, isValid}] must contain a closing </mark> tag');
    }

    invalidCriteria = newCriteriaList.find(({ message }) => message.includes('<script>'));
    if (invalidCriteria) {
      throw new Error("message in [{message, isValid}] mustn't contain a <script> tag");
    }
  }

  render() {
    const content = this.criteriaList?.map(({ message, isValid }) => {
      let validationClass = '';

      if (isValid !== undefined) {
        validationClass = isValid ? ' valid' : ' invalid';
      }

      return (
        <div class={`criteria${validationClass}`}>
          <div class="icon" />
          <span class="message" innerHTML={message}></span>
        </div>
      );
    });

    return <Host>{content}</Host>;
  }
}
