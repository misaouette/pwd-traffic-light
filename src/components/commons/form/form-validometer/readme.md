# form-validometer



<!-- Auto Generated Below -->


## Properties

| Property                    | Attribute      | Description                                                                                                                                                                                                                                | Type         | Default     |
| --------------------------- | -------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ | ------------ | ----------- |
| `criteriaList` _(required)_ | --             | Array of {isValid, message} objects describing the valid and invalid criteria. - `message` must contain <mark> element if isValid is defined. - leave `isValid` undefined if the message is purely descriptive and contains no validation. | `Criteria[]` | `undefined` |
| `vspaceSmall`               | `vspace-small` | Flag to add small vertical space above the validometer                                                                                                                                                                                     | `boolean`    | `undefined` |


## Dependencies

### Used by

 - [form-input](../form-input)

### Graph
```mermaid
graph TD;
  form-input --> form-validometer
  style form-validometer fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
