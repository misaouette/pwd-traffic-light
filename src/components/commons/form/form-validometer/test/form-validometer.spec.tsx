import { newSpecPage } from '@stencil/core/testing';
import { FormValidometer } from '../form-validometer';

describe('form-validometer', () => {
  it("throws when message doesn't include entering <mark> tag", async () => {
    const options = {
      components: [FormValidometer],
      html: `<form-validometer criteriaList={[{message: 'url must start with http://</mark>', isValid: true},]}></form-validometer>`,
    };
    //   expect(await newSpecPage(options)).toThrow();
    await expect(newSpecPage(options)).rejects.toEqual('message in [{message, isValid}] must contain an opening <mark> tag : (url must start with http://</mark>)');
  });

  it("throws when message doesn't include entering <mark> tag", async () => {
    const options = {
      components: [FormValidometer],
      html: `<form-validometer criteriaList={[{message: 'url <mark>must start with http://', isValid: true},]}></form-validometer>`,
    };
    //    expect(await newSpecPage(options)).toThrow();
    await expect(newSpecPage(options)).rejects.toEqual('message in [{message, isValid}] must contain a closing </mark> tag : (url must start with http://</mark>)');
  });

  it('renders a list of valid and invalid criteria', async () => {
    const page = await newSpecPage({
      components: [FormValidometer],
      html: `<form-validometer criteriaList={[{message: 'url <mark> must start with http://</mark>', isValid: true},{message: 'url <mark> cannot contain localhost</mark>', isValid: false}]}></form-validometer>`,
    });
    expect(page.root).toMatchSnapshot();
  });
});
