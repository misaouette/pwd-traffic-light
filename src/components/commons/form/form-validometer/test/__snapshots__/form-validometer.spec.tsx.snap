// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`form-validometer renders a list of valid and invalid criteria 1`] = `
<form-validometer 'url="" <mark="" criterialist="{[{message:">
  <mock:shadow-root></mock:shadow-root>
  must start with http://', isValid: true},{message: 'url
  <mark>
    cannot contain localhost
  </mark>
  ', isValid: false}]}&gt;
</form-validometer>
`;
